File: backend/src/domain/entities/Subject.ts
export class Subject {
  constructor(
    public readonly id: string,
    public readonly userId: string,
    public readonly semesterId: string,
    public name: string,
    public color: string,
    public icon: string | undefined,
    public isActive: boolean,
    public readonly createdAt: Date,
    public updatedAt: Date
  ) {}

  updateDetails(name: string, color: string, icon?: string): void {
    this.name = name;
    this.color = color;
    this.icon = icon;
    this.updatedAt = new Date();
  }

  updateName(name: string): void {
    this.name = name;
    this.updatedAt = new Date();
  }

  updateColor(color: string): void {
    this.color = color;
    this.updatedAt = new Date();
  }

  updateIcon(icon: string): void {
    this.icon = icon;
    this.updatedAt = new Date();
  }

  deactivate(): void {
    this.isActive = false;
    this.updatedAt = new Date();
  }

  activate(): void {
    this.isActive = true;
    this.updatedAt = new Date();
  }

  belongsToUser(userId: string): boolean {
    return this.userId === userId;
  }

  static create(
    id: string,
    userId: string,
    semesterId: string,
    name: string,
    color: string,
    icon?: string
  ): Subject {
    return new Subject(
      id,
      userId,
      semesterId,
      name,
      color,
      icon,
      true,
      new Date(),
      new Date()
    );
  }
}

File: mobile/lib/src/domain/models/subject/subject.dart
class Subject {
  Subject({
    required this.id,
    required this.semesterId,
    required this.name,
    required this.color,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.icon,
  });

  factory Subject.fromJson(Map<String, dynamic> json) => Subject(
        id: json['id'] as String,
        semesterId: json['semesterId'] as String,
        name: json['name'] as String,
        color: json['color'] as String,
        icon: json['icon'] as String?,
        isActive: json['isActive'] as bool? ?? true,
        createdAt: DateTime.parse(json['createdAt'] as String).toLocal(),
        updatedAt: DateTime.parse(json['updatedAt'] as String).toLocal(),
      );

  final String id;
  final String semesterId;
  final String name;
  final String color;
  final String? icon;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
}
